### ISSUE TITLE
Add new scriptlet that prevents some anti-debugging

### ISSUE THREAD

**Issue by D4niloMR** [uBlockOrigin/uBlock-issues#3378]
### Prerequisites

- [X] I verified that this is not a filter list issue. Report any issues with filter lists or broken website functionality in the [uAssets issue tracker](https://github.com/uBlockOrigin/uAssets/issues).
- [X] This is **NOT** a [_YouTube_](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [_Facebook_](https://www.github.com/uBlockOrigin/uAssets/issues/3367) or [_Twitch_](https://www.github.com/uBlockOrigin/uAssets/issues/5184) report. _These sites MUST be reported by clicking their **respective links**_.
- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

Mostly for websites that has obfuscated code with anti-debugging. Maybe in the future it will be possible to replace the snippet below with [trusted-replace-argument ](https://github.com/gorhill/uBlock/wiki/Resources-Library#trusted-replace-argumentjs-).

Only the proxy is necessary.

```js
window.Function.prototype.constructor = new Proxy(window.Function.prototype.constructor, {
  apply: function(target, thisArg, args) {
    if (typeof args[0] === "string" && args[0].includes("debugger")) {
      return;
    }
    return Reflect.apply(target, thisArg, args);
    }
  });
window.console.clear = function () {};
Object.defineProperties(Object.prototype, { devtoolsDetector:{}, Debugger:{} });
```

### A specific URL where the issue occurs.

```yaml
`https://veev.to/d/4vz566d3awg0`
`https://www.japscan.lol/`
`https://animet2.net/`
`https://binged.to/watch/movie/365177`
```

### Steps to Reproduce

1. Disable annoyances lists.
2. Choose one of the URLs above.
3. Navigate to the site and open devtools.
4. Boom, debugger statement

### Expected behavior

Visiting site with obfuscated code the debugger statement is not executed.

### Actual behavior

Visiting site with obfuscated code and anti debugger, the debugger statement is executed.

### uBO version

1.59.1b22

### Browser name and version

Firefox Developer Edition 131.0b6

### Operating System and version

Windows 10

**Comment by D4niloMR** [comment]
Doesn't work, I don't see nothing logged with `##+js(trusted-replace-argument, Function.prototype.constructor)` either.

**Comment by gorhill** [comment]
There is this extension specialized in defusing anti-devtools: https://github.com/Andrews54757/Anti-Anti-Debug

AMO: https://addons.mozilla.org/en-US/firefox/addon/anti-anti-debug/
CWS: https://chromewebstore.google.com/detail/anti-anti-debug/mnmnmcmdkigakhlfkcdimghndnmomfeo

**Comment by gorhill** [comment]
`trusted-replace-argument` should work but doesn't because the trapped target ends up being `construct`, which is problematic for `Function` since it can be called either through `new Fcuntion(...)` or directly through `Function(...)`. Also, if I end up fixing this somehow, the function argument can be a different position than 0, so the scriptlet might need a way to provide argument position relative to the end, and/or a way to scan all arguments.

**Comment by gorhill** [comment]
`trusted-suppress-native-method` uses the same internal helper to trap, so probably the same issue.

**Comment by gorhill** [comment]
`##+js(trusted-replace-argument, Function.prototype.constructor, 0, , condition, debugger)` works in 1.59.1rc0.

**Comment by gorhill** [comment]
Why not try it?

**Comment by D4niloMR** [comment]
It works, but seems to take a lot of resources, especially on Firefox.

<details><summary>Screenshots</summary>

![image](https://github.com/user-attachments/assets/49b366ea-5c3e-4782-95e2-109f0d73e8f9)

![image](https://github.com/user-attachments/assets/646e2214-afbf-43a9-b5f5-5ec12040c20b)

</details> 

https://share.firefox.dev/3ZoT6tD

**Comment by gorhill** [comment]
> seems to take a lot of resources, especially on Firefox

This happens without the filter too, just opening the dev console?

**Comment by D4niloMR** [comment]
Yes, but I wasn't with devtools opened while profiling and it still used a lot of resources.

I can't reproduce this issue anymore on that site though. I can on `japscan.lol`.

**Comment by gorhill** [comment]
Looks like it's related to the scriptlet sending logging information to uBO's background process -- might be an issue when a site does something very fast non-stop which is being trapped by uBO.

**Comment by garry-ut99** [comment]
I can reproduce, whole browser begun freezing: 

<details>
<summary>screenshoot</summary>

![ss](https://github.com/user-attachments/assets/62fe91cf-5245-438b-91f0-de695e96fa15)

</details>

**Comment by D4niloMR** [comment]
Now it becomes unresponsive, more noticeable when the logger/devtools is open. Firefox still freezes.

**Comment by gorhill** [comment]
I am going to need all the details of what I need to do to reproduce.

I could reproduce performance issue only when the logger was opened when navigating to `https://veev.to/d/4vz566d3awg0` while using the filter `veev.to##+js(trusted-replace-argument, Function.prototype.constructor, 0, , condition, debugger)`, and this was squarely a issue with sending too much logging information to the logger from the page context -- as the page probably keep creating `debugger` functions non-stop in a tight loop. Once I added the throttling code to prevent repeatedly reporting thesame information to the logger, the issue went away. The profiling data you shared also show the issue was the logging overhead.

Now if you have another scenario causing issue, I will need all the details on how to reproduce on my side, and profiling data is welcome.

**Comment by D4niloMR** [comment]
The issue of the resources is fixed. Now I'm reporting that on my end, navigating through the devtools tabs is unresponsive, like network tab, sources/debugger and expanding nodes in inspector.

But this is probably the site's own issue, I didn't notice that the snippet in the description and the Anti Anti Debug extension was throwing error and stopping the debugger loop. Once I added `veev.to##+js(trusted-replace-argument, Function.prototype.constructor, 0, noopFunc, condition, debugger)` the issue went away.

**Comment by gorhill** [comment]
> `veev.to##+js(trusted-replace-argument, Function.prototype.constructor, 0, noopFunc, condition, debugger)`

You shouldn't be using `noopFunc`, the argument is supposed to be a string representing the body of the function to create and using `noopFunc` causes an exception -- try `Function.prototype.constructor(function(){})` in dev console of `https://example.com/`).

This doesn't throw:

    veev.to##+js(trusted-replace-argument, Function.prototype.constructor, 0, /* debugger */, condition, debugger)

You can try at `https://example.com/` console: `Function.prototype.constructor('/* debugger */')`

**Comment by D4niloMR** [comment]
Yes, but on my end if it doesn't throw I suffer from the issue above.

**Comment by garry-ut99** [comment]
I do confirm what D4niloMR is saying: 
- when using `noopFunc` in the scriptlet:
   - dev tools work fine: are responsive and switching between dev tools tabs is responsive as well 
   - and logger says: `Replaced argument: Before: "debugger" After: function(){}` 
- when using `/* debugger */` or any other random string value:
   - dev tools are unresponsive / laggy (like without the scriptlet)
   - and logger says: `Replaced argument: Before: "debugger" After: undefined`

<hr>

> You shouldn't be using `noopFunc`, the argument is supposed to be a string representing the body of the function to create

Description says: "-- the value is interpreted the same way the value for `set-constant` is interpreted." - I guess D4niloMR interpreted it that it should be fed with the same values like from `set-constant` scriptlet, hence that's why he used `noopFunc`, and it actually does work (for him and for me)...


> and using `noopFunc` causes an exception -- try `Function.prototype.constructor(function(){})` in dev console of `https://example.com/`).

For me it throws  after pasted manually in dev console, but not when used in the scriptlet on the site, maybe that's why it works for me and for D4niloMR.

**Comment by stephenhawk8054** [comment]
How about this:

```adb
veev.to##+js(trusted-replace-argument, Function.prototype.constructor, 0, {”value”: "/ * debugger * /"}, condition, debugger)
```

**Comment by garry-ut99** [comment]
Logger says: `Replaced argument: Before: "debugger" After: ` ("works in")
Logger says: `Replaced argument: Before: "debugger" After: /* debugger */` ("How about this")
Logger says: `Replaced argument: Before: "debugger" After: undefined` ("so need someone to test")

In all cases, dev tools are still unresponsive / laggy / freezing (like without the scriptlet) every time dev tools are opened or closed. The only filter which works and fixes dev tools, is the one with `noopFunc`.

**Comment by gorhill** [comment]
> Yes, but on my end if it doesn't throw I suffer from the issue above.

Right, I didn't realize an exception worked better. In that case, this filter might be a better match:

    veev.to##+js(trusted-suppress-native-method, Function.prototype.constructor, '"debugger"', abort)

**Comment by garry-ut99** [comment]
The above filter works fine as well for me (on both Firefox and Chromium).
